File: LarkCardKit/Builders/SelectPersonBuilder.cs
using LarkCardKit.Models.Behaviors;
using LarkCardKit.Models.Elements;

namespace LarkCardKit.Builders;

public class SelectPersonBuilder
{
    private readonly SelectPerson _selectPerson = new();
    
    public SelectPersonBuilder Name(string name)
    {
        _selectPerson.Name = name;
        return this;
    }
    
    public SelectPersonBuilder Required(bool required = true)
    {
        _selectPerson.Required = required;
        return this;
    }
    
    public SelectPersonBuilder Placeholder(string placeholder)
    {
        _selectPerson.Placeholder = new PlainText { Content = placeholder };
        return this;
    }
    
    public SelectPersonBuilder InitialOption(string openId)
    {
        _selectPerson.InitialOption = openId;
        return this;
    }
    
    public SelectPersonBuilder Width(string width)
    {
        _selectPerson.Width = width;
        return this;
    }
    
    public SelectPersonBuilder Disabled(bool disabled = true)
    {
        _selectPerson.Disabled = disabled;
        return this;
    }
    
    public SelectPersonBuilder ElementId(string id)
    {
        _selectPerson.ElementId = id;
        return this;
    }
    
    public SelectPersonBuilder Margin(string margin)
    {
        _selectPerson.Margin = margin;
        return this;
    }
    
    public SelectPersonBuilder Confirm(string title, string text)
    {
        _selectPerson.Confirm = new ConfirmConfig
        {
            Title = new PlainText { Content = title },
            Text = new PlainText { Content = text }
        };
        return this;
    }
    
    public SelectPersonBuilder OnChange(object callbackData)
    {
        _selectPerson.Behaviors ??= new List<object>();
        _selectPerson.Behaviors.Add(new CallbackBehavior { Value = callbackData });
        return this;
    }
    
    public SelectPerson Build() => _selectPerson;
}

File: TypeScript/LarkCardKit/src/builders/SelectPersonBuilder.ts
import type { SelectPerson, MultiSelectPerson, PersonSelectedOption } from '../models/elements';

export class SelectPersonBuilder {
  private _select: SelectPerson = { tag: 'select_person', name: '' };

  name(name: string): this {
    this._select.name = name;
    return this;
  }

  placeholder(placeholder: string): this {
    this._select.placeholder = { tag: 'plain_text', content: placeholder };
    return this;
  }

  selectedOption(userId: string, userName?: string, avatarUrl?: string): this {
    this._select.selected_options ??= [];
    const opt: PersonSelectedOption = { user_id: userId };
    if (userName) opt.user_name = userName;
    if (avatarUrl) opt.avatar_url = avatarUrl;
    this._select.selected_options.push(opt);
    return this;
  }

  multiple(multiple: boolean = true): this {
    this._select.multiple = multiple;
    return this;
  }

  disabled(disabled: boolean = true): this {
    this._select.disabled = disabled;
    return this;
  }

  allowedUserIds(userIds: string[]): this {
    this._select.allowed_user_ids = userIds;
    return this;
  }

  elementId(id: string): this {
    this._select.element_id = id;
    return this;
  }

  margin(margin: string): this {
    this._select.margin = margin;
    return this;
  }

  build(): SelectPerson {
    return this._select;
  }
}

export class MultiSelectPersonBuilder {
  private _select: MultiSelectPerson = { tag: 'multi_select_person', name: '' };

  name(name: string): this {
    this._select.name = name;
    return this;
  }

  placeholder(placeholder: string): this {
    this._select.placeholder = { tag: 'plain_text', content: placeholder };
    return this;
  }

  selectedOption(userId: string, userName?: string, avatarUrl?: string): this {
    this._select.selected_options ??= [];
    const opt: PersonSelectedOption = { user_id: userId };
    if (userName) opt.user_name = userName;
    if (avatarUrl) opt.avatar_url = avatarUrl;
    this._select.selected_options.push(opt);
    return this;
  }

  disabled(disabled: boolean = true): this {
    this._select.disabled = disabled;
    return this;
  }

  allowedUserIds(userIds: string[]): this {
    this._select.allowed_user_ids = userIds;
    return this;
  }

  elementId(id: string): this {
    this._select.element_id = id;
    return this;
  }

  margin(margin: string): this {
    this._select.margin = margin;
    return this;
  }

  build(): MultiSelectPerson {
    return this._select;
  }
}
